src/analysis/bounded_variation.lean

```diff
@@ -634,6 +635,147 @@ lemma monotone_on.has_locally_bounded_variation_on {f : α → ℝ} {s : set α}
   has_locally_bounded_variation_on f s :=
 λ a b as bs, ((hf.evariation_on_le as bs).trans_lt ennreal.of_real_lt_top).ne
 
+section variation_on_from_to
+
+
+/--
+The **signed** variation of `f` on the interval `Icc a b` intersected with the set `s`,
+squashed to a real (therefore only really meaningful if the variation is finite)
+-/
+noncomputable def variation_on_from_to (f : α → E) (s : set α) (a b : α) : real :=
+if a ≤ b then (evariation_on f (s ∩ Icc a b)).to_real else
+            - (evariation_on f (s ∩ Icc b a)).to_real
+
+variables {f : α → E} {s : set α}
+
+lemma variation_on_from_to_self (a : α) : variation_on_from_to f s a a = 0 :=
+begin
+  dsimp only [variation_on_from_to],
+  rw [if_pos (le_refl _), Icc_self, evariation_on.subsingleton, ennreal.zero_to_real],
+  exact λ x hx y hy, hx.2.trans hy.2.symm,
+end
+
+lemma variation_on_from_to_nonneg_of_le {a b : α} (h : a ≤ b) : 0 ≤ variation_on_from_to f s a b :=
+by { dsimp only [variation_on_from_to], simp only [if_pos h, ennreal.to_real_nonneg], }
+
+lemma variation_on_from_to_nonpos_of_ge {a b : α} (h : b ≤ a) : variation_on_from_to f s a b ≤ 0 :=
+begin
+  dsimp only [variation_on_from_to],
+  split_ifs with h',
+  { cases le_antisymm h h',
+    rw [Icc_self, evariation_on.subsingleton, ennreal.zero_to_real],
+    exact λ x hx y hy, hx.2.trans hy.2.symm, },
+  { simp only [right.neg_nonpos_iff, ennreal.to_real_nonneg], },
+end
+
+lemma variation_on_from_to_eq_of_le {a b : α} (h : a ≤ b) :
+  variation_on_from_to f s a b = (evariation_on f (s ∩ Icc a b)).to_real := if_pos h
+
+lemma variation_on_from_to_eq_of_ge {a b : α} (h : b ≤ a) :
+  variation_on_from_to f s a b = - (evariation_on f (s ∩ Icc b a)).to_real :=
+begin
+  dsimp only [variation_on_from_to],
+  split_ifs with h',
+  { cases le_antisymm h h',
+    rw [Icc_self, evariation_on.subsingleton, ennreal.zero_to_real],
+    simp only [neg_zero],
+    exact λ x hx y hy, hx.2.trans hy.2.symm, },
+  { refl, }
+end
+
+lemma variation_on_from_to_eq_neg_swap (a b : α) :
+  variation_on_from_to f s a b = - variation_on_from_to f s b a :=
+begin
+  rcases lt_trichotomy a b with ab|rfl|ba,
+  { simp only [variation_on_from_to, if_pos ab.le, if_neg ab.not_le, neg_neg], },
+  { simp only [variation_on_from_to_self, neg_zero], },
+  { simp only [variation_on_from_to, if_pos ba.le, if_neg ba.not_le, neg_neg], },
+end
+
+lemma variation_on_from_to_add (hf : has_locally_bounded_variation_on f s)
+  {a b c : α} (ha : a ∈ s) (hb : b ∈ s) (hc : c ∈ s) :
+  variation_on_from_to f s a b + variation_on_from_to f s b c = variation_on_from_to f s a c :=
+begin
+  rcases le_total a b with ab|ba;
+  rcases le_total a c with ac|ca;
+  rcases le_total b c with bc|cb,
```

Comment:
This simple lemma would reduce you to one case (would require mathlib4 PR though). Anyone aware of other proofs that may be golfed using this?
```
import algebra.group.basic

variables {α β : Type*} [group β] (f : α → α → β)
-- no need of commutativity

@[to_additive additive_of_total]
lemma multiplicative_of_total [linear_order α]
  (hswap : ∀ a b, f a b = (f b a)⁻¹)
  (hmul : ∀ {a b c}, a ≤ b → b ≤ c → f a c = f a b * f b c)
  (a b c : α) : f a c = f a b * f b c :=
begin
  have := _,
  obtain hbc | hcb := le_total b c,
  { revert b c, exact this },
  { simp [this c b hcb, hswap c b] },
  intros b c hbc,
  obtain hab | hba := le_total a b,
  { exact hmul hab hbc },
  obtain hac | hca := le_total a c,
  { simp [hmul hba hac, hswap a b] },
  { simp [hswap a b, hmul hbc hca, hswap a c] },
end
```

Reply:
Very neat! I think it's worth it, and I can open the mathlib4 PR, but where would you put that lemma?

Reply:
Huh, let me take that back: the mathlib4 sync process is kind of a hassle. Maybe we can just have a TODO note somewhere to extract the lemma when ported?

Reply:
Does it become harder to sync recently? I think the requirement to update the SHA in the ported file (after the mathlib3 PR is merged) is maybe a recent addition, but isn't hard to do. In the meantime I've come up with a version with generalized hypotheses [monoid] + [is_total], and will PR it to *algebra.group.basic* to mathlib3&4 if you don't want to.
```
import algebra.group.basic

variables {α β : Type*} [monoid β] (f : α → α → β)

@[to_additive additive_of_total]
lemma multiplicative_of_total (r : α → α → Prop) [t : is_total α r]
  (hswap : ∀ a b, f a b * f b a = 1)
  (hmul : ∀ {a b c}, r a b → r b c → f a c = f a b * f b c)
  (a b c : α) : f a c = f a b * f b c :=
begin
  have := _,
  obtain hbc | hcb := t.total b c,
  { revert b c, exact this },
  { rw [this c b hcb, mul_assoc, hswap c b, mul_one] },
  intros b c hbc,
  obtain hab | hba := t.total a b,
  { exact hmul hab hbc },
  obtain hac | hca := t.total a c,
  { rw [hmul hba hac, ← mul_assoc, hswap a b, one_mul] },
  { rw [← one_mul (f a c), ← hswap a b, hmul hbc hca, mul_assoc, mul_assoc, hswap c a, mul_one] },
end
```

Reply:
Does it become harder to sync recently?

Not really afaiu, but I always seem to be doing something wrong, and I sort of feel like I should do less of these post-sync PRs, but I won't complain if you take care of it!

Reply:
Following @urkud's suggestion I've now generalized `multiplicative_of_total` to take into account a symmetric binary predicate such as integrability or rectifiability, so this lemma now also applies to `ereal` (which forms an `add_comm_monoid`) when the variations involved are finite.

Reply:
Do you propose making `evariation_on_from_to` ereal-valued then?